Drop unused django-taggit dependency

relrod opened the pull request:
##### SUMMARY

This drops the django-taggit dependency and drops the relevant fields from old migrations.

<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

relrod committed:
Drop unused django-taggit dependency

This drops the django-taggit dependency and drops the relevant fields
from old migrations.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod commented:
Refs #14238

But it goes further than the initial ask in that ticket, insofar as it actually removes the dep and references to it.

Note that this does not add a migration - because any migration will want these fields to exist, which of course they won't after this patch.

This might not be an issue given (per #14238):
> Manually delete the taggit tables at the start of the migration command or some other way, or do this as SQL in another numbered migration

But really I am just fiddling around at 3am and I am probably not thinking straight.

matburt commented:
👍🏻 This stuff goes back a *long* time... love to see it removed.

AlanCoding commented:
This is a draft, so it's not expected to be complete yet, but I wanted to round up the known items that are still TODO.

 - Introduce a data migration, and inside of that
   - Drop the tables used by the taggit app (I believe there are two)
   - Delete content types that reference the taggit app, `ContentType.objects.filter(app_label='taggit')`
   - [more speculative] delete the migration records that reference the app `MigrationRecorder.Migration.objects.filter(app='taggit').delete()`

This migration may be reversible in a superficial sense, as there will essentially be 2 forms of history. There is the upgrade scenario where the user starts with the app installed, and has it removed through the data migration. The other scenario is a new install and in the case the app is never installed. So reversing will not re-install the app, and we could allow reversing it under the 2nd form of history.

relrod committed:
Add migration to drop old taggit data

- Drop the taggit tables themselves
- Delete the contenttypes
- Delete the previous migration records

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod commented:
@AlanCoding Sorry for the delay here, finally circling back to this. I added a migration, I think it is how we talked about doing this; it addresses the 3 points in your comment.

relrod committed:
Just use a string here

Signed-off-by: Rick Elrod <rick@elrod.me>

AlanCoding commented on awx/main/migrations/0186_drop_django_taggit.py:
```diff
@@ -0,0 +1,29 @@
+
+
+def delete_taggit_migration_records(apps, schema_editor):
+    from django.db.migrations.recorder import MigrationRecorder
```
Since you've already imported the module at the top, you can get rid of this line and reference `migrations.recorder.MigrationRecorder`. Sometimes you can't do this, but I tested it with python shell and it works.

AlanCoding commented on awx/main/migrations/0186_drop_django_taggit.py:
```diff
@@ -0,0 +1,29 @@
+
+    operations = [
+        migrations.RunSQL("DROP TABLE IF EXISTS taggit_tag CASCADE;"),
+        migrations.RunSQL("DROP TABLE IF EXISTS taggit_taggeditem CASCADE;"),
```
:+1:  nicely done. This is as close to supported as we're ever going to get.

relrod committed:
remove superfluous import

Signed-off-by: Rick Elrod <rick@elrod.me>

AlanCoding commented:
I like this patch, and I want to get it in. So I took some time to front-run some of the tests and give independent verification. This is upstream testing, using the docker-compose environment. Nevertheless, I believe it obtains the necessary coverage.

 - :heavy_check_mark: Upgrade from `devel` to this branch with old data, confirm jobs and stuff still work
 - :heavy_check_mark: Create a fresh database and migrate to this PR. Confirmed accounts and jobs still work.

After the migrations, as expected the taggit app no longer appears in the showmigrations output.
